File: go/hashtable/hashtable.go
package main

import (
	"fmt"
	"hash/fnv"
)

// should be bigger and prime
// but I did less for testing
const tableSize = 10

type Node struct {
	key   string
	value interface{}
	next  *Node
}

type HashTable struct {
	table [tableSize]*Node
}

func NewHashTable() *HashTable {
	return &HashTable{}
}

func (ht *HashTable) hash(key string) uint32 {
	h := fnv.New32a()
	h.Write([]byte(key))
	return h.Sum32() % tableSize
}

func (ht *HashTable) Put(key string, value interface{}) {
	index := ht.hash(key)
	if ht.table[index] == nil {
		ht.table[index] = &Node{key: key, value: value}
		return
	}

	current := ht.table[index]
	for current != nil {
		if current.key == key {
			current.value = value
			return
		}
		if current.next == nil {
			current.next = &Node{key: key, value: value}
			return
		}
		current = current.next
	}
}

func (ht *HashTable) Get(key string) (interface{}, bool) {
	index := ht.hash(key)
	current := ht.table[index]

	for current != nil {
		if current.key == key {
			return current.value, true
		}
		current = current.next
	}
	return nil, false
}

func (ht *HashTable) Remove(key string) bool {
	index := ht.hash(key)
	if ht.table[index] == nil {
		return false
	}

	if ht.table[index].key == key {
		ht.table[index] = ht.table[index].next
		return true
	}

	current := ht.table[index]
	for current.next != nil {
		if current.next.key == key {
			current.next = current.next.next
			return true
		}
		current = current.next
	}
	return false
}

func (ht *HashTable) String() string {
	result := ""
	for i := 0; i < tableSize; i++ {
		result += fmt.Sprintf("Bucket[%d]: ", i)
		current := ht.table[i]
		for current != nil {
			result += fmt.Sprintf("(%s:%v) ", current.key, current.value)
			current = current.next
		}
		result += "\n"
	}
	return result
}

File: rust/hashtable/src/hashtable.rs
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

// should be prime number
// should be mutch bigger but is this way for testing
const TABLE_SIZE: usize = 10;

#[derive(Debug)]
struct Entry<K, V> {
    key: K,
    value: V,
}

pub struct HashTable<K, V> {
    table: Vec<Vec<Entry<K, V>>>,
}

impl<K: Hash + Eq, V> HashTable<K, V> {
    pub fn new() -> Self {
        let mut table = Vec::with_capacity(TABLE_SIZE);
        for _ in 0..TABLE_SIZE {
            table.push(Vec::new());
        }
        HashTable { table }
    }

    fn hash(&self, key: &K) -> usize {
        let mut hasher = DefaultHasher::new();
        key.hash(&mut hasher);
        hasher.finish() as usize % TABLE_SIZE
    }

    pub fn put(&mut self, key: K, value: V) {
        let index = self.hash(&key);
        let bucket = &mut self.table[index];

        for entry in bucket.iter_mut() {
            if entry.key == key {
                entry.value = value;
                return;
            }
        }
        bucket.push(Entry { key, value });
    }

    pub fn get(&self, key: &K) -> Option<&V> {
        let index = self.hash(key);
        let bucket = &self.table[index];

        bucket
            .iter()
            .find(|entry| entry.key == *key)
            .map(|entry| &entry.value)
    }

    pub fn remove(&mut self, key: &K) -> bool {
        let index = self.hash(key);
        let bucket = &mut self.table[index];

        if let Some(pos) = bucket.iter().position(|entry| entry.key == *key) {
            bucket.remove(pos);
            return true;
        }
        false
    }
}

impl<K, V> std::fmt::Display for HashTable<K, V>
where
    K: std::fmt::Debug,
    V: std::fmt::Debug,
{
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        for (i, bucket) in self.table.iter().enumerate() {
            writeln!(f, "Bucket[{}]: {:?}", i, 
                bucket.iter().map(|entry| 
                    format!("({:?}:{:?})", entry.key, entry.value)
                ).collect::<Vec<_>>().join(" ")
            )?;
        }
        Ok(())
    }
}